### DISCUSSION TITLE
Typing component that uses "as" prop

### DISCUSSION THREAD

**Discussion by gtiminskis-sc** [ariakit/ariakit#1358] [Questions] [answered]
Hello (again)! 
I'm a bit of a Typescript newbie so this question might be a bit of a silly one.
I'm trying to create a wrapper around the Tab component and allow it to be a link:

```
// MyTab.tsx
import { Tab as ReakitTab, TabProps as ReakitTabProps } from "reakit/Tab";
type TabProps = ReakitTabProps;

// Can do either with or without forwardRef here
const MyTab = forwardRef<any, TabProps>((props: TabProps, ref) => {
  const { disabled, ...rest } = props;
  return (
    <ReakitTab ref={ref} disabled={disabled} focusable={!disabled} {...rest} />
  );
});
```

```
// Usage:
import { Link } from "react-router-dom";
import { MyTab, useTabState } from "my-component-lib/tabs";

const tab = useTabState();
<MyTab
  {...tab}
  as={Link}
  to="https://google.com"
>
  My Tab as Link
</MyTab>
```

The issue is that the `as` property throws a Type error 
```
Property 'as' does not exist on type 'IntrinsicAttributes & Pick<PropsWithChildren<Pick<TabProps, "baseId" | "unstable_idCountRef" | "unstable_virtual" | "orientation" | "unstable_moves" | "unstable_hasActiveWidget" | ... 22 more ... | "wrapElement"> & RefAttributes<...> & Omit<...>>, "baseId" | ... 27 more ... | "wrapElement"> & RefAttributes<...>'.
```

===========================================================================
I've also tried implementing it like this:
```
export type PropsOf<
  C extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
> = JSX.LibraryManagedAttributes<C, React.ComponentPropsWithoutRef<C>>;

type TabProps<T extends ElementType> = ReakitTabProps & {
  as: T;
} & PropsOf<T>;
```
```
const MyTab = <T extends ElementType>(props: TabProps<T>) { ... }
```
but this makes `MyTab` accept any prop and seems to lose the typings for both `Link` and `ReakitTabProps`.

Is there any way for my wrapper component to accept a React Element prop such as `Link` and automatically pick up its types?

(Currently using "reakit": "^1.3.11")

**Comment by diegohaz** [comment, marked as answer]
Typing the `as` prop is complicated. I'd recommend you avoid exposing this API if you can. If you do need it, though, here's a CodeSandbox that may help: https://codesandbox.io/s/typescript-as-prop-dicj8?file=/src/App.tsx
